mlrun/model_monitoring/applications/base.py

```diff
@@ -0,0 +1,273 @@
+# Copyright 2023 Iguazio
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from abc import ABC, abstractmethod
+from typing import Any, Union, cast
+
+import numpy as np
+import pandas as pd
+
+import mlrun
+import mlrun.model_monitoring.applications.context as mm_context
+import mlrun.model_monitoring.applications.results as mm_results
+from mlrun.serving.utils import StepToDict
+
+
+class ModelMonitoringApplicationBaseV2(StepToDict, ABC):
+    """
+    A base class for a model monitoring application.
+    Inherit from this class to create a custom model monitoring application.
+
+    example for very simple custom application::
+        # mlrun: start-code
+        class MyApp(ApplicationBase):
+            def do_tracking(
+                self,
+                monitoring_context: mm_context.MonitoringApplicationContext,
+            ) -> ModelMonitoringApplicationResult:
+                monitoring_context.log_artifact(
+                    TableArtifact(
+                        "sample_df_stats", df=self.dict_to_histogram(sample_df_stats)
+                    )
+                )
+                return ModelMonitoringApplicationResult(
+                    name="data_drift_test",
+                    value=0.5,
+                    kind=mm_constant.ResultKindApp.data_drift,
+                    status=mm_constant.ResultStatusApp.detected,
+                )
+
+
+        # mlrun: end-code
+    """
+
+    kind = "monitoring_application"
+
+    def do(
+        self, monitoring_context: mm_context.MonitoringApplicationContext
+    ) -> tuple[
+        list[
+            Union[
+                mm_results.ModelMonitoringApplicationResult,
+                mm_results.ModelMonitoringApplicationMetric,
+            ]
+        ],
+        mm_context.MonitoringApplicationContext,
+    ]:
+        """
+        Process the monitoring event and return application results & metrics.
+        :param monitoring_context:   (MonitoringApplicationContext) The monitoring application context.
+        :returns:                    (list[ModelMonitoringApplicationResult], dict) The application results
```

Comment:
just a recommendation (not mandatory) - when the return value is a tuple it is nicer to explain in with the following format: 
```suggestion
        :returns:  A tuple of:
             [0] = list of application results that can be either from type `ModelMonitoringApplicationResult` or from type `ModelMonitoringApplicationResult`.
             [1] = the original application event, wrapped in `MonitoringApplicationContext` object
```

Reply:
+ note that in your docstrings the return value is` list[ModelMonitoringApplicationResult], dict` while in the function signature is actually 
```
tuple[
        list[
            Union[
                mm_results.ModelMonitoringApplicationResult,
                mm_results.ModelMonitoringApplicationMetric,
            ]
        ],
        mm_context.MonitoringApplicationContext,
    ]:
```